### PR TITLE
Add "kubectl validate" command to do a cluster health check.

### PR DESCRIPTION
**Pull request by fabioy**
This is the first step to resolving issue #5946.

### CODE REVIEW

**Comment by satnam6502** [comment]
I think Zach is in a better position to review this PR.

File: pkg/client/client.go [hunk 1]
```diff
@@ -137,6 +138,25 @@ func (c *Client) ServerAPIVersions() (*api.APIVersions, error) {
 	return &v, nil
 }
 
+type ClusterValidationStateInterface interface {
+	ClusterValidationState() (*api.ServerStatusList, error)
+}
+
+// ClusterValidationState retrieves and parses the server's self-monitored cluster state.
+func (c *Client) ClusterValidationState() (*api.ServerStatusList, error) {
+	body, err := c.Get().AbsPath("/validate").Do().Raw()
```
Comment (roberthbailey) [inline pkg/client/client.go]: Add a TODO to make this hit a versioned validate endpoint?

File: pkg/client/client.go [hunk 2]
```diff
@@ -137,6 +138,25 @@ func (c *Client) ServerAPIVersions() (*api.APIVersions, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	statuses := []api.ServerStatus{}
+	err = json.Unmarshal(body, &statuses)
```
Comment (roberthbailey) [inline pkg/client/client.go]: if err := json.Unmarshal(...); err != nil {
}

File: pkg/kubectl/cmd/validate.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
```
Comment (roberthbailey) [inline pkg/kubectl/cmd/validate.go]: 2015

File: pkg/kubectl/cmd/validate_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	f, tf, codec := NewAPIFactory()
+	tf.Printer = &testPrinter{}
+	tf.Client = &client.FakeRESTClient{
+		Codec: codec,
+		Client: client.HTTPClientFunc(func(req *http.Request) (*http.Response, error) {
+			switch p, m := req.URL.Path, req.Method; {
+			case p == "/validate" && m == "GET":
+				body, err := json.Marshal(testStatusList.Items)
+				if err != nil {
+					t.Fatalf("unexpected json marshal error: %v", err)
```
Comment (roberthbailey) [inline pkg/kubectl/cmd/validate_test.go]: strings passed to log functions should start with a capital letter

File: pkg/kubectl/resource_printer.go [hunk 5]
```diff
@@ -631,6 +634,21 @@ func printResourceQuotaList(list *api.ResourceQuotaList, w io.Writer) error {
...
 }
 
+func printServerStatus(item *api.ServerStatus, w io.Writer) error {
+	_, err := fmt.Fprintf(w, "%s\t%s\t%d\t%s\t%s\n", item.Component, item.Health, item.HealthCode, item.Msg, item.Err)
+	return err
+}
+
+func printServerStatusList(list *api.ServerStatusList, w io.Writer) error {
+	for _, item := range list.Items {
+		if err := printServerStatus(&item, w); err != nil {
```
Comment (roberthbailey) [inline pkg/kubectl/resource_printer.go]: It might be better not to short-circuit here but to print all of the status lines and then return an error at the end if any individual line caused an error. That way if there is more than one error you can see all of them at once.

File: pkg/client/client.go (hunk 1 above)
Comment (satnam6502) [inline pkg/client/client.go]: Does this return an `api.Status`? If not then `DoRaw()` might be better.

**Comment by roberthbailey** [comment]
Since this makes changes to the api and @zmerlynn is out this week, I'm going to bounce this over to @lavalamp for server-side review. @jlowdermilk might also want to take a look at the kubectl changes.

File: pkg/client/client.go (hunk 1 above)
Comment (roberthbailey) [inline pkg/client/client.go]: Is DoRaw() different from Do().Raw()?

Comment (satnam6502) [inline pkg/client/client.go]: Yes, I discovered a bug where `Do().Raw()` does the wrong thing when the non-`apiStatus` JSON response includes a `status` field as is the case with Elasticsearch. I refactored it into two distinct phases which first processes the response without assuming it is an `api.Status` (this is done by `DoRaw()`) and then interpret it as an `api.Status` which is done by `Do()` which now calls `DoRaw()`. I think we should replace all occurrences of `Do().Raw()` with `DoRaw()` and remove the `Raw()` method.

Comment (fabioy) [inline pkg/client/client.go]: Added TODO and changed to DoRaw().

File: pkg/client/client.go (hunk 2 above)
Comment (fabioy) [inline pkg/client/client.go]: Done.

File: pkg/kubectl/cmd/validate.go (hunk 3 above)
Comment (fabioy) [inline pkg/kubectl/cmd/validate.go]: Done.

File: pkg/kubectl/cmd/validate_test.go (hunk 4 above)
Comment (fabioy) [inline pkg/kubectl/cmd/validate_test.go]: Done.

File: pkg/kubectl/resource_printer.go (hunk 5 above)
Comment (fabioy) [inline pkg/kubectl/resource_printer.go]: This was the pattern for the other resource printers, which I followed. Also, an error in the "printServerStatus" call would generally mean that there was an error in Fprintf, which means the data returned didn't quite match the expected types. Continuing to attempt to print the other lines seems...redundant (the given line probably wasn't printed as well).

A final argument against doing this is that the user may be able to get at the data by requesting an output of YAML or JSON.

**Comment by fabioy** [comment]
Rebased and pushed the changes. PTAL. Thank you.

File: pkg/api/serialization_test.go [hunk 6]
```diff
@@ -128,7 +128,7 @@ func TestList(t *testing.T) {
 	roundTripSame(t, item)
 }
 
-var nonRoundTrippableTypes = util.NewStringSet("ContainerManifest", "ContainerManifestList")
+var nonRoundTrippableTypes = util.NewStringSet("ContainerManifest", "ContainerManifestList", "ServerStatus")
```
Comment (lavalamp) [inline pkg/api/serialization_test.go]: New objects should never appear in this list. You have a bug in your conversions if you need this.

File: pkg/api/v1beta3/types.go [hunk 7]
```diff
@@ -1609,3 +1610,19 @@ type SecretList struct {
 
 	Items []Secret `json:"items" description:"items is a list of secret objects"`
 }
+
+// ServerStatus (and ServerStatusList) holds the cluster validation info.
+type ServerStatus struct {
+	Component  string       `json:"component,omitempty"`
+	Health     string       `json:"health,omitempty"`
+	HealthCode probe.Result `json:"healthCode,omitempty"`
```
Comment (lavalamp) [inline pkg/api/v1beta3/types.go]: probe.Result probably shouldn't be used here (it can stay in the unversioned types) because then a change to it would break the api.

File: pkg/client/client.go [hunk 8]
```diff
@@ -43,6 +43,7 @@ type Interface interface {
 	NamespacesInterface
 	PersistentVolumesInterface
 	PersistentVolumeClaimsNamespacer
+	ClusterValidationStateInterface
```
Comment (lavalamp) [inline pkg/client/client.go]: Why not just ClusterValidator as a name.

File: pkg/client/client.go [hunk 9]
```diff
@@ -137,6 +138,25 @@ func (c *Client) ServerAPIVersions() (*api.APIVersions, error) {
 	return &v, nil
 }
 
+type ClusterValidationStateInterface interface {
+	ClusterValidationState() (*api.ServerStatusList, error)
+}
+
+// ClusterValidationState retrieves and parses the server's self-monitored cluster state.
+// TODO: This should hit the versioned endpoint when that is implemented.
+func (c *Client) ClusterValidationState() (*api.ServerStatusList, error) {
```
Comment (lavalamp) [inline pkg/client/client.go]: ValidateCluster() as a name?

**Comment by lavalamp** [comment]
I have pointed out some problems. Reassigning to @bgrant0607 to approve an API change.

File: pkg/kubectl/cmd/cmd.go [hunk 10]
```diff
@@ -65,6 +65,7 @@ Find more information at https://github.com/GoogleCloudPlatform/kubernetes.`,
 	cmds.AddCommand(NewCmdClusterInfo(f, out))
 	cmds.AddCommand(NewCmdApiVersions(f, out))
 	cmds.AddCommand(NewCmdVersion(f, out))
+	cmds.AddCommand(NewCmdValidateCluster(f, out))
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/cmd.go]: We have `cluster-info`. How about `cluster-validate`? `validate` is just too generic, and we're likely to want to use that for validating user objects.

**Comment by bgrant0607** [comment]
More context please? The PR description and cited issue say almost nothing.

**Comment by bgrant0607** [comment]
This is going to be a breaking API change. Have you asked on kubernetes-dev whether anyone is using the current /validate API?

File: pkg/api/types.go [hunk 11]
```diff
@@ -1767,3 +1768,19 @@ func AddToNodeAddresses(addresses *[]NodeAddress, addAddresses ...NodeAddress) {
 		}
 	}
 }
+
+// ServerStatus (and ServerStatusList) holds the cluster validation info.
+type ServerStatus struct {
```
Comment (bgrant0607) [inline pkg/api/types.go]: Please remove the json tags here. This is the internal rep.

File: pkg/api/v1beta3/types.go [hunk 12]
```diff
@@ -1609,3 +1610,19 @@ type SecretList struct {
 
 	Items []Secret `json:"items" description:"items is a list of secret objects"`
 }
+
+// ServerStatus (and ServerStatusList) holds the cluster validation info.
+type ServerStatus struct {
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: This should be called ComponentStatus.

File: pkg/api/v1beta3/types.go [hunk 13]
```diff
@@ -1609,3 +1610,19 @@ type SecretList struct {
...
+	Component  string       `json:"component,omitempty"`
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: Name.

File: pkg/api/v1beta3/types.go [hunk 14]
```diff
@@ -1609,3 +1610,19 @@ type SecretList struct {
...
+	Health     string       `json:"health,omitempty"`
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: What is this field populated with? All fields must bear description tags, which will show up in swagger.

File: pkg/api/v1beta3/types.go [hunk 15]
```diff
@@ -1609,3 +1610,19 @@ type SecretList struct {
...
+	HealthCode probe.Result `json:"healthCode,omitempty"`
+	Msg        string       `json:"msg,omitempty"`
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: No abbreviations in the API. Message.

File: pkg/api/v1beta3/types.go [hunk 16]
```diff
@@ -1609,3 +1610,19 @@ type SecretList struct {
...
+	Err        string       `json:"err,omitempty"`
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: Did you consider embedding Status instead? There should at least be a compelling reason why this is significantly different than Status.

File: pkg/apiserver/validator.go [hunk 17]
```diff
@@ -109,7 +102,7 @@ func (v *validator) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		} else {
 			errorMsg = "nil"
 		}
-		reply = append(reply, ServerStatus{name, status.String(), status, msg, errorMsg})
+		reply = append(reply, api.ServerStatus{name, status.String(), status, msg, errorMsg})
```
Comment (bgrant0607) [inline pkg/apiserver/validator.go]: That's not the versioned object.

It seems like a good idea to version this API, but it needs to change pretty significantly to make that possible.

For instance, we need to install version-specific handlers at version-specific paths.

The current InstallValidator:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/apiserver/apiserver.go#L143

probably needs to move to:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/apiserver/api_installer.go

However, since Openshift also uses apiserver and Kubelet will use it in the future, it can't just be installed by default. 

This may be the first example where we really need a separate API prefix #3806, with separate types.go files #1744.

A simpler approach for now would be to not version this API yet:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/unversioned.go

That would prohibit breaking changes post-1.0, but we could expose new resources, if necessary.

We should rename this one to /componentstatuses, if that's what it is.

Also, this needs to use go-restful, so that the documentation shows up in Swagger.

cc @smarterclayton, in case he has preferences/requirements or suggestions.

File: pkg/kubectl/cmd/cmd.go (hunk 10 above)
Comment (bgrant0607) [inline pkg/kubectl/cmd/cmd.go]: I take it back. If this is just component status, we should call it `component-status`.

If we versioned the APIs, it could be:
```
kubectl get cs
```

**Comment by fabioy** [comment]
To give a bit more context, the "/validate" call does effectively the same thing that "validate-cluster.sh" does, so I looked into rewriting the later to just call it. This led to "This should be callable from kubectl". Which this tries to do. I'll change the names and fix the code so that the type is in the unversioned type (I was trying to set it up so that it could be versioned in the future, but I guess this is not the right way to do it).

As far as the end point, my idea is to keep the "/validate" end point for monitoring and other tools, but also have versioned API for kubectl and others. The "/validate" could then just be an alias to the latest version.

**Comment by bgrant0607** [comment]
/validate is unversioned. It's not acceptable to break compatibility, so it can't be an alias to the latest version. Either version the API, or don't change it in a way that would ever break compatibility.

**Comment by bgrant0607** [comment]
It looks like test/e2e/networking.go is the only thing really using `/validate` today? If so, then we should convert it to a versioned API now, before too many dependencies crop up and make it unchangeable.

**Comment by fabioy** [comment]
I'd flip this around and say that if you want to avoid breaking changes, then use the versioned API. The "/validate" endpoint should have a "best effort" backwards compat, which is why being an alias to the latest version seems ok. This it a diagnostic endpoint and shouldn't be overburdened.

BTW, I believe GKE uses the "/validate" end point in its verification.

**Comment by roberthbailey** [comment]
I can confirm that GKE currently relies on /validate.

**Comment by bgrant0607** [comment]
The fact that GKE relies on /validate should be setting off warning bells. 

There is an increasing amount of cluster turnup code that's not in our github repo: Tectonic, Murano, Magnum (Heat), Stratos, ... Adding support to kubectl will further increase the visibility of /validate. Assuming it can be changed without breaking lots of people seems dubious. 

We're trying to minimize the number of unversioned APIs to the bare essentials, such as discovering what api prefixes and what api versions exist.

The existence of this unversioned pseudo-API is unfortunate.

**Comment by smarterclayton** [comment]
How would a 3rd party extend this cluster status check?  Is it expected that services that sit around Kube have their own /validate?

**Comment by fabioy** [comment]
cc @brendandburns 

I believe the intent of "/validate" was to have a single place where k8s cluster validation logic can go. Currently it does health ping to minions and master processes (i.e. etcd, kube-scheduler, etc). I don't think it's the proper place for provider specific checks (I'll leave it to @brendandburns to comment further on this, since he wrote it). In my view, this data would be used along with other provider-specific stuff to get at the full cluster health picture.

Exposing it via kubectl just makes it a bit nicer to get at the data. Fundamentally, it's just a diagnostic data dump (though I can see the value of also making it versioned).

**Comment by bgrant0607** [comment]
I don't mind having a "diagnostic" endpoint, but then it should go under /debug and nothing should depend on it.

**Comment by fabioy** [comment]
When you say "nothing should depend on it", do you mean it shouldn't be callable via kubectl?

Also, perhaps a different way to think of it is that "/validate" is a centralized "/healthz" for the cluster (with richer info).

**Comment by fabioy** [comment]
On further thought, how about this:

1. Create a top level kubectl command called "debug" or "diagnose". We can put the proposed "validate" under it: "kubectl debug components-state" (or something like that).
2. The "debug" command is unversioned and "best effort" backwards compat. 
3. This would also enable future additions (i.e. "debug network" or "debug service etcd").

This still leaves open the question of whether "/validate" should exist as is, or be a subpath under "/debug". I'd prefer to keep it a top-level path, like the health ping "/healthz", since it was intended to be used in a similar way (for an easy cluster health ping).

**Comment by bgrant0607** [comment]
By "nothing should depend on it" I mean that if it were removed, it would just affect human-oriented UX. No systems would break.

**Comment by bgrant0607** [comment]
The path should go under /debug, or it should be a real API.

**Comment by fabioy** [comment]
It seems like the best way forward is to simply version this call (issue #6658). Circling back to one of the feedback items, I figure I'll add:

kubectl get component-status
kubectl get component-status <component name>

The one oddity is that a "get" request could trigger work on the master (i.e. ping nodes, etc). Issue #6047 mentions adding a caching layer to this call. I'll also move "validate-cluster.sh" to use this.

Regardless of the changes proposed, we'll have to keep the "/validate" end point for a while (until GKE moves off it). I think I can refactor the code so that it's not a dup.

**Comment by bgrant0607** [comment]
Making it a versioned API SGTM.

On kubectl syntax, please copy replicationcontrollers:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/kubectl.go#L100

The resource name should not use camelcase nor dashes, and should be plural.

I don't think it makes sense to include nodes, since we already have an API for nodes.

I would pattern the API on NodeCondition:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta3/types.go#L1129

We're growing a similar API for PodCondition:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta3/types.go#L732

The structs could be unified once all the same information is reported.

**Comment by yujuhong** [comment]
To clarify, issue #6047 is not to add a caching layer, but to avoid querying all nodes. Nodes already post their statuses to apiserver regularly, so it'd be easy for apiserver to return the latest status instead of talking to each node.

@bgrant0607, is your suggestion to remove node statuses completely from the /validate endpoint? If so, I think #6047 should be folded into this issue since there is no need to optimize node status anymore.

**Comment by bgrant0607** [comment]
@yujuhong Yes, exactly, remove node statuses from /validate.

**Comment by fabioy** [comment]
That'd imply that a "validate cluster" call would comprised of multiple
kubectl calls and some aggregation. The nice thing of "/validate" is that
it includes it all, which is good for monitoring.

**Comment by fabioy** [comment]
***This is still a WIP***

Pushed a change that has the versioned object, "componentstatus". Please take a look and let me know if this seems acceptable. Still working on tests and some cleanup. Will push the final changes later.

Thanks.

**Comment by zmerlynn** [comment]
Can you summarize what you ended up with, if it's not a single api call? I'm either "somewhat" to "very" against this if it's not a single call.

**Comment by bgrant0607** [comment]
@zmerlynn What do you plan to do with the call?

**Comment by zmerlynn** [comment]
I think I misunderstood the proposals. If the difference is between whether
there is a difference between a single call for "master components" versus
"node components" versus a single call for each, I'm not sure I care. If,
however, we're forced to loop over individual nodes, I would like to force
the API to do that, because, as I've tried to make clear in other
conversations, GKE is a very dumb client.

**Comment by fabioy** [comment]
Added tests and cleanup. PTAL.

**Comment by fabioy** [comment]
@bgrant0607 Ping. Does the object model seem ok?

**Comment by bgrant0607** [comment]
@fabioy Will take a look tonight. In the future, feel free to ping me by direct email. I'm drowning in Github notifications.

File: pkg/registry/componentstatus/rest.go [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func (rs *REST) NewList() runtime.Object {
+	return &api.ComponentStatusList{}
+}
+
+// Returns the list of component status. Note that the label and field are both ignored.
+// Note that this call doesn't support labels or selectors.
+func (rs *REST) List(ctx api.Context, label labels.Selector, field fields.Selector) (runtime.Object, error) {
+	servers := rs.GetServersToValidate()
```
Comment (bgrant0607) [inline pkg/registry/componentstatus/rest.go]: Eventually I want the apiserver to be a dumb, CRUDy server. Eventually, a controller would need to implement all the business logic and post status back. 

We'll need a registration mechanism for components. Right now, we sort of have a fixed set of components on one node, but that's about to change Real Soon Now. Already we have "add ons" that run on the cluster but which are required, or will be (DNS), and full API plugins are planned. The components are all being converted to static pods as we speak. Others are working on HA, so components will be replicated. Full self-hosting of those components is the eventual goal.

What should this return if there are 5 etcd instances, 3 apiservers, 2 schedulers, and 2 controller managers?

Comment (bgrant0607) [inline pkg/registry/componentstatus/rest.go]: I'm wishing we had the experimental API prefix discussed in #6363 and elsewhere.

File: pkg/apiserver/validator_test.go [hunk 19]
```diff
@@ -65,8 +65,8 @@ func TestValidate(t *testing.T) {
 				StatusCode: test.code,
 			},
 		}
-		fake := &http.Client{Transport: fakeRT}
-		status, data, err := s.check(fake)
+		//		fake := &http.Client{Transport: fakeRT}
```
Comment (bgrant0607) [inline pkg/apiserver/validator_test.go]: Delete or uncomment.

File: pkg/api/v1beta3/types.go [hunk 20]
```diff
@@ -1641,3 +1641,34 @@ type SecretList struct {
...
+// Type and constants for component health validation.
+type ComponentConditionType string
+
+// These are the valid conditions for the component.
+const (
+	ComponentHealthy ComponentConditionType = "Healthy"
+)
+
+type ComponentCondition struct {
+	Type    ComponentConditionType `json:"type"`
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: All non-inline fields need description tags.

File: pkg/api/v1beta3/types.go [hunk 21]
```diff
@@ -1641,3 +1641,34 @@ type SecretList struct {
...
+	Message string                 `json:"message,omitempty" description:"health check message received from the component"`
+	Error   string                 `json:"error,omitempty" description:"error code from health check attempt (if any)"`
+}
+
+// ComponentStatus (and ComponentStatusList) holds the cluster validation info.
+type ComponentStatus struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	Name       string               `json:"name,omitempty" description:"name of the component"`
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: ObjectMeta includes a Name field already. It might as well match the name of the component.

File: pkg/api/v1beta3/types.go [hunk 22]
```diff
@@ -1641,3 +1641,34 @@ type SecretList struct {
...
+
+type ComponentCondition struct {
+	Type    ComponentConditionType `json:"type"`
+	Status  ConditionStatus        `json:"status"`
+	Message string                 `json:"message,omitempty" description:"health check message received from the component"`
+	Error   string                 `json:"error,omitempty" description:"error code from health check attempt (if any)"`
+}
+
+// ComponentStatus (and ComponentStatusList) holds the cluster validation info.
+type ComponentStatus struct {
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: Given where we're headed, I think we're going to need a full-blown Component object, with Spec as well as Status -- the same pattern as Pod and the other objects. The /components API would act as the component registry.

If we assume we're going to run all the components in pods, then we can use existing Pod mechanisms to perform liveness and readiness probes, but there's still the issue of how to collect the results of the probes. 

Potentially we'd also define services (possibly headless) for components, especially if replicated. We've discussed copying readiness condition status into Endpoints. We could maybe directly use that info, or mimic the pattern and copy the conditions from selected pods into the ComponentStatus.

We could go ahead with what you've done, but I'm concerned it won't actually last very long. If we did go ahead with this, I'd consider ComponentStatus a "virtual resource" -- it's not actually persisted in etcd. Therefore, we probably could use ListMeta in place of ObjectMeta, in which case the custom Name field could remain.

It's probably worth a VC to discuss before you do more work on this.

Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: Also, FYI, we're going to discuss HA at this week's community hangout. This functionality seems like a good issue to raise.

File: pkg/registry/componentstatus/rest.go (hunk 18 above)
Comment (fabioy) [inline pkg/registry/componentstatus/rest.go]: "What should this return if there are 5 etcd instances, 3 apiservers, 2 schedulers, and 2 controller managers?"

The call is to the current master api server, which tries it best to return the status of the currently active components. It seems irrelevant to list the "stand-by" components. When HA become available, this "extended" reporting could be done under an option flag.

File: pkg/apiserver/validator_test.go (hunk 19 above)
Comment (fabioy) [inline pkg/apiserver/validator_test.go]: I'll remove it.

File: pkg/api/v1beta3/types.go (hunk 22 above)
Comment (fabioy) [inline pkg/api/v1beta3/types.go]: I think having a Component object eventually is fine. ComponentStatus is also very much intended to be a virtual resource -- it shouldn't be persisted anywhere. The trick is that who gets to be called a component and how it gets registered. Right now the list is a fixed list, with fixed end points it hits.

File: pkg/api/v1beta3/types.go (hunk 21 above)
Comment (fabioy) [inline pkg/api/v1beta3/types.go]: Removed the extra "Name" field.

File: pkg/api/v1beta3/types.go (hunk 20 above)
Comment (fabioy) [inline pkg/api/v1beta3/types.go]: Done.

**Comment by fabioy** [comment]
Addressed comments, sync'ed, rebased and pushed changes. PTAL. Thank you!

**Comment by brendandburns** [comment]
LGTM.

This PR has gone on long enough.  This looks good enough to merge as a first round (and its certainly an improvement over what was there previously)

To be honest, blocking this PR on making it into a real API seems kind of like Overkill to me, but I was out, and now its done so merge and move on...

**Comment by bgrant0607** [comment]
AFAICT, this PR had unresolved feedback. Maybe changes weren't pushed?

**Comment by bgrant0607** [comment]
It also demonstrates another bug was introduced into verify-description.sh, since this should be failing it:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta1/types.go#L1606

File: pkg/api/types.go [hunk 23]
```diff
@@ -1827,3 +1827,34 @@ func AddToNodeAddresses(addresses *[]NodeAddress, addAddresses ...NodeAddress) {
...
+type ComponentConditionType string
+
+// These are the valid conditions for the component.
+const (
+	ComponentHealthy ComponentConditionType = "Healthy"
+)
+
+type ComponentCondition struct {
+	Type    ComponentConditionType `json:"type" description:"the type of condition"`
+	Status  ConditionStatus        `json:"status" description:"the status of this condition"`
```
Comment (bgrant0607) [inline pkg/api/types.go]: This file shouldn't have description tags. Description tags need to be on all non-inline fields in the versioned types.go files.

File: pkg/api/v1beta2/types.go [hunk 24]
```diff
@@ -1655,3 +1655,32 @@ type SecretList struct {
...
+type ComponentConditionType string
+
+// These are the valid conditions for the component.
+const (
+	ComponentHealthy ComponentConditionType = "Healthy"
+)
+
+type ComponentCondition struct {
+	Type    ComponentConditionType `json:"type"`
+	Status  ConditionStatus        `json:"status"`
```
Comment (bgrant0607) [inline pkg/api/v1beta2/types.go]: Missing description tags.

File: pkg/api/v1beta1/types.go [hunk 25]
```diff
@@ -1592,3 +1592,32 @@ type GlusterfsVolumeSource struct {
...
+type ComponentConditionType string
+
+// These are the valid conditions for the component.
+const (
+	ComponentHealthy ComponentConditionType = "Healthy"
+)
+
+type ComponentCondition struct {
+	Type    ComponentConditionType `json:"type"`
+	Status  ConditionStatus        `json:"status"`
```
Comment (bgrant0607) [inline pkg/api/v1beta1/types.go]: Missing description tags.

File: pkg/kubectl/kubectl.go [hunk 26]
```diff
@@ -96,17 +96,19 @@ func (e ShortcutExpander) VersionAndKindForResource(resource string) (defaultVer
...
 func expandResourceShortcut(resource string) string {
 	shortForms := map[string]string{
-		"po": "pods",
-		"rc": "replicationcontrollers",
-		// DEPRECATED: will be removed before 1.0
-		"se":     "services",
-		"svc":    "services",
-		"mi":     "minions",
+		// Please keep this alphabetized
+		"cs":     "componentstatus",
```
Comment (bgrant0607) [inline pkg/kubectl/kubectl.go]: should have been plural

File: pkg/api/types.go (hunk 23 above)
Comment (fabioy) [inline pkg/api/types.go]: Interesting, I have the standard presubmit hooks. Filed issue #7080 for the "verify-description.sh" error. I'll prepare a separate PR and address these.

File: pkg/api/v1beta2/types.go (hunk 24 above)
Comment (fabioy) [inline pkg/api/v1beta2/types.go]: Added them.

File: pkg/api/v1beta1/types.go (hunk 25 above)
Comment (fabioy) [inline pkg/api/v1beta1/types.go]: Added them.

File: pkg/kubectl/kubectl.go (hunk 26 above)
Comment (fabioy) [inline pkg/kubectl/kubectl.go]: Changed to "componentstatuses".

**Comment by fabioy** [comment]
I've created a separate PR to resolve these issues. PR #7082. Thank you.

**Comment by YitongFeng** [comment]
Is this feature still work? how to use kubectl validate? In k8s 1.9.9, kubectl validate command not found
